src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -126,16 +128,51 @@ internal sealed class AddDocumentsAction(
             ImmutableArray<DocumentState> documents)
             : TranslationAction(oldProjectState, newProjectState)
         {
+            /// <summary>
+            /// Amount to break batches of documents into.  That allows us to process things in parallel, without also
+            /// creating too many individual actions that then need to be processed.
+            /// </summary>
+            public const int AddDocumentsBatchSize = 32;
+
             public readonly ImmutableArray<DocumentState> Documents = documents;
 
             public override async Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
             {
-                // Parse all the documents in parallel.
-                using var _ = ArrayBuilder<Task<SyntaxTree>>.GetInstance(this.Documents.Length, out var tasks);
+                using var _1 = ArrayBuilder<SyntaxTree>.GetInstance(this.Documents.Length, out var trees);
+
+#if NETSTANDARD
+
+                // Parallel parse the documents we have in chunks.
+                using var _2 = ArrayBuilder<Task<SyntaxTree>>.GetInstance(AddDocumentsBatchSize, out var tasks);
+
+                // Can simplify this to use span-slicing once the language supports using ref-structs in async methods.
+                for (int batchStart = 0, total = this.Documents.Length; batchStart < total; batchStart += AddDocumentsBatchSize)
+                {
+                    tasks.Clear();
+
+                    for (int i = batchStart, batchEnd = Math.Min(batchStart + AddDocumentsBatchSize, total); i < batchEnd; i++)
+                    {
+                        var document = this.Documents[i];
```

Comment:
>var document = this.Documents[i]; [](http://example.com/codeflow?start=24&length=33)

One particularly slow GetSyntaxTreeAsync call here could completely block any new work from starting. What about using a semaphore to do the parallelism? Something like:

```
var throttler = new SemaphoreSlim(initialCount: AddDocumentsBatchSize);
var tasks = this.Documents.Select(async document =>
{
    await throttler.WaitAsync(cancellationToken).ConfigureAwait(false);
    try
    {
        return await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
    }
    finally
    {
        throttler.Release();
    }
});

var batch = await Task.WhenAll(tasks).ConfigureAwait(false);
trees.AddRange(batch);
```